### PR TITLE
ARROW-6375: [C++] Extend ConversionTraits to allow efficiently appending list values in STL API

### PR DESCRIPTION
**Pull request by ozars**
`ConversionTraits<T>` is extended to allow users to provide `AppendMultipleRows` method for a type. This method can be used for constructing `List` or `LargeList` more efficiently by calling value builder's `ApendValues()` method with a contiguous range (e.g. `std::vector`), instead of iterating over each element  with value builder's `Append()` method. Default is to fallback to the latter behavior if `AppendMultipleRows` isn't provided.

### CODE REVIEW

File: cpp/src/arrow/stl.h [hunk 1]
```diff
@@ -50,19 +52,33 @@ using BareTupleElement = typename std::remove_const<typename std::remove_referen
...
-    constexpr static bool nullable = false;                                         \
-  };
+#define ARROW_STL_CONVERSION(CType_, ArrowType_)                                         \
+  template <>                                                                            \
+  struct ConversionTraits<CType_> : public CTypeTraits<CType_> {                         \
+    static Status AppendRow(typename TypeTraits<ArrowType_>::BuilderType& builder,       \
+                            CType_ cell) {                                               \
+      return builder.Append(cell);                                                       \
+    }                                                                                    \
+    template <typename Range>                                                            \
```
Comment (bkietz) [inline cpp/src/arrow/stl.h]: I don't think it's appropriate to make this a member of the ConversionTraits struct unless you standardize on std::vector for your cell range concept; it makes adding new implementations very verbose and detecting whether the functionality is available is now impossible at compile time:
```c++
static_assert(
    HasAppendMultipleRows<bool, BooleanBuilder, MemoryPool*>::value,
    "memory pool shouldn't be usable as a boolean cell range");
```

Please make this a non member method or disallow non-vector cell ranges

File: cpp/src/arrow/stl.h [hunk 2]
```diff
@@ -87,33 +103,88 @@ struct ConversionTraits<std::string> : public CTypeTraits<std::string> {
...
+    ARROW_RETURN_NOT_OK(builder.Append());
+    auto& value_builder =
+        ::arrow::internal::checked_cast<ElementBuilderType<ValueCType>&>(
+            *builder.value_builder());
+    return ConversionTraits<ValueCType>::AppendMultipleRows(value_builder, cell_range);
+  }
+};
+}  // namespace internal
+
+/// Add multiple rows to a (large) list builder.
```
Comment (bkietz) [inline cpp/src/arrow/stl.h]: This is used to append a single value to a list builder; it appends multiple values to the child builder. Please update the comment and rename this function accordingly

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/stl.h (hunk 1 above)
Comment (ozars) [reply cpp/src/arrow/stl.h]: Yes, that's a good point. How about using an enabler:
`template <typename Range, typename = enable_if<is_same<decay<Range>, std::vector<CType_>>>`

This addresses HasAppendMultipleRows issue. Could you explain a bit more what you mean by adding new implementations?

File: cpp/src/arrow/stl.h (hunk 2 above)
Comment (ozars) [reply cpp/src/arrow/stl.h]: Will do. I named it to be consistent with AppendRow, as it also appends to child builder in current implementation. How about `AppendMultipleValues`?

File: cpp/src/arrow/stl.h (hunk 1 above)
Comment (ozars) [reply cpp/src/arrow/stl.h]: I think I got what you meant. I will try putting this functionality as a free function.

File: cpp/src/arrow/stl.h (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/stl.h]: Since this is list specific, maybe AppendListValue?

Comment (ozars) [reply cpp/src/arrow/stl.h]: Renamed this helper function as `AppendCellRange` and free function provided for specializaton as `AppendListValues`. Does it sound right?

**Review by bkietz — APPROVED** [review]
LGTM, will merge on passing build

**Pull request closed without merging**